daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Util.scala

```diff
@@ -100,4 +104,112 @@ object Util {
       case None => Right(None)
     }
 
+  /** Compute the inputs to a DAML transaction (in traversal order), that is, the referenced contracts, parties,
+    *  and key.
+    */
+  def collectInputs(
+      tx: Transaction.Transaction,
+      addContract: ContractId => Unit,
+      addParty: Ref.Party => Unit,
+      addGlobalKey: GlobalKey => Unit,
+  ): Unit = {
+
+    val localContract = tx.localContracts
+
+    def addContractInput(coid: ContractId): Unit =
+      if (!localContract.isDefinedAt(coid))
+        addContract(coid)
+
+    def addPartyInputs(parties: Set[Ref.Party]): Unit =
+      parties.toList.sorted(Ref.Party.ordering).foreach(addParty)
```

Comment:
Why do you have to sort the parties?

Reply:
KV needs it to get deterministic results.

Reply:
I drop the sort.